### PR TITLE
Replace the concept of virtual block size, introduce the concept of block cost, and express CBS as a function of it

### PR DESCRIPTION
**Pull request by NicolasDorier**
The concept of MAX_BLOCK_SIZE can be removed more easily for the next block size HF.
Using MAX_BLOCK_COST make it easier to adapt HF proposals after segwit deployment, because the limited resources will not really be MAX_BLOCK_SIZE anymore.

This will have another advantages for later: The notion of "Block cost" can be later tweaked to include other parameters like SigOp cost.

We can imagine, in the next hardfork, to remove the limitation on the number of SigOps, and just include SigOps as an additional cost parameter in the "Block Cost".

Current rule about sig op could even be coded into GetBlockCost: (pseudocode)
```
size_t GetBlockCost(const CBlock& block)
{
    // witness data is 4 times less costly than core data, because we can prune it
    size_t coreCost = ::GetSerializeSize(block, SER_NETWORK, 0);
    size_t witCost = ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS) - coreCost;
    size_t sigCost = GetSigOpCount(block) > MAX_SIG_OPS ? MAX_BLOCK_COST + 1 : 0;
    return witCost + 4 * coreCost + sigCost;
}
```

### CODE REVIEW

File: src/primitives/block.cpp [hunk 1]
```diff
@@ -32,11 +32,10 @@ std::string CBlock::ToString() const
...
+size_t GetBlockCost(const CBlock& block)
 {
-    // The formula is: vsize = base_size + witness_size / 4.
-    // We can only serialize base or totalbase+witness, however, so the formula
-    // becomes: vsize = base_size + (total_size - base_size) / 4 or
-    // vsize = (total_size + 3 * base_size) / 4.
-    return (::GetSerializeSize(block, SER_NETWORK, 0) * 3 + ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS) + 3) / 4;
+    // witness data is 4 times less costly than core data, because we can prune it
+    size_t coreCost = ::GetSerializeSize(block, SER_NETWORK, 0);
+    size_t witCost = ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS) - coreCost;
```
Comment (CodeShark) [inline src/primitives/block.cpp]: Subtractions like this make me a little uneasy...unless you bound the result at zero. Although perhaps it's overkill here, assuming SERIALIZE_TRANSACTION_WITNESS always produces a larger result than 0 - but the problem is that a bug in the serialization code could lead to underflow issues which could lead to extremely bizarre behavior.

**Comment by CodeShark** [comment]
Actually, an assert would be best

File: src/primitives/block.cpp (hunk 1 above)
Comment (CodeShark) [inline src/primitives/block.cpp]: size_t witCost = ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS);
assert(witCost >= coreCost);
witCost -= coreCost;

File: src/primitives/block.cpp [hunk 2]
```diff
@@ -32,11 +32,13 @@ std::string CBlock::ToString() const
...
-    // becomes: vsize = base_size + (total_size - base_size) / 4 or
-    // vsize = (total_size + 3 * base_size) / 4.
-    return (::GetSerializeSize(block, SER_NETWORK, 0) * 3 + ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS) + 3) / 4;
+size_t GetBlockCost(const CBlock& block)
+{    
+    // coreSize is the size of block without witness data
+    size_t coreSize = ::GetSerializeSize(block, SER_NETWORK, 0);
+    // witSize is the size of witness data: totalBlockSize - coreSize
+    size_t witSize = ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS) - coreSize;
+    assert(witSize >= 0);
```
Comment (CodeShark) [inline src/primitives/block.cpp]: size_t is an unsigned integer.
do:
size_t witSize = ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS);
assert(witSize >= coreSize);
witSize -= coreSize;

Comment (sipa) [inline src/primitives/block.cpp]: How about just using my original formula which has no subtractions?

Comment (CodeShark) [inline src/primitives/block.cpp]: Yes, just multiply it through by 4.

**Comment by NicolasDorier** [comment]
@CodeShark I made your changes.

@sipa your formula was for calculating the "Virtual Block Size". A concept that I'd like to get away to introduce the concept of "Block cost" which is what should be used as a base for fee calculation for example for a transaction for example.

If you prefer using the virtual block size, I see no objection to that, but I think I detailed nice reason to have a block cost metric instead.

**Comment by CodeShark** [comment]
ACK on the terminology. I'll let sipa speak to the formula clarification.

**Comment by NicolasDorier** [comment]
@sipa disregard my previous comment I misunderstood what you meant.
What you propose is:
```
return coreSize * 3 + totalSize;
```
I'm not really for it, it is less easy to review. But I can change the PR if you insist.

**Comment by NicolasDorier** [comment]
@sipa @CodeShark I updated the code to get rid of the substraction after sleeping on it. :)

**Comment by sipa** [comment]
My only concern here is things like feerate values that are stored/communicated, as they will become "per kilocost" instead of "per kilobyte".

**Comment by NicolasDorier** [comment]
As they should be. Since this is the real limiting factor now. And "per kilobyte" is technically true as long as the cost only include segwit and core size.
If you are interested into this PR, I can put it on segwit3.

**Comment by sipa** [comment]
@NicolasDorier Right, but we want backward compatibility. So when for example a command line option for feerates is given, which used to be intended as per kilobyte, we'd likely want to multiply the passed value by 4 to set the internal per-cost variable.

**Comment by NicolasDorier** [comment]
Ok I understand what you mean now. This PR does not break the meaning of feerate though. Feerate is still calculated on the size, not on the cost. This is a problem unrelated to this PR.

But indeed worth noting: As miners must maximize the fee/cost to maximize profit, consumers of command line option or RPC will still be dealing with fee/bytes.
This is not an urgent thing to fix, but the concept of feerate will need to be deprecated anyway, since this is not aligned with miner's incentive anymore.

**Comment by sipa** [comment]
You can't calculate fee based on size. The optimal strategy for miners is
to choose transactions sorted by fee per cost, as cost is the limited
resource now. I believe my segnet branches already do that.

**Comment by NicolasDorier** [comment]
yes, I agree. This PR does not change anything about that.
Do you want I send this PR on segwit3 ?

**Pull request closed without merging**